lib/tether/tether.go

```diff
@@ -600,20 +625,53 @@ func (t *tether) Start() error {
 func (t *tether) Stop() error {
 	defer trace.End(trace.Begin(""))
 
-	// cancel the context to signal waiters
-	t.cancel()
+	// cancel the context to signal waiters and indicate shutdown
+	t.reloadCancel()
+
+	// trigger a reload - this should deliver the stopsignal to the children
+	t.reload <- struct{}{}
 
-	// TODO: kill all the children
 	close(t.reload)
 
+	// this gets called from inside HandleSessionExit so cannot take session locks
+	// TODO: add a mechanism to block for clean shutdown of:
+	//   session processes
+	//   extensions
+	// probably should wait for the long overdue tether rewrite
+	// instead we're going to rely on the outstanding pid count via tether.Wait
+	timeout, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	err := t.Wait(timeout)
+	cancel()
+
+	t.cancel()
+
+	return err
+}
+
+func (t *tether) Wait(ctx context.Context) error {
+	tick := time.Tick(500 * time.Millisecond)
+
+	pids := t.lenChildPid()
+	for pids > 0 && ctx.Err() == nil {
+		log.Infof("Waiting for %d processes to exit", pids)
+		pids = t.lenChildPid()
+		<-tick
```

Comment:
I would do it this way:
```
time.Sleep(500 * time.Millisecond)
```

Reply:
This way makes it regular without adding on time spend blocking in `lenChildPid()`